DOCS add secure application access draft (#2450)
truenas/documentation: content/SCALE/SCALETutorials/Apps/AppSecurity/_index.md

If you want to access your application via subdomain (shown in this guide) two environment variables must be set in the Nextcloud application.
[overwrite.cli.url](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwrite-cli-url) and [overwritehost](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwritehost)

Enter the two environment variables in **Name** as *OVERWRITECLIURL* and *OVERWRITEHOST*.

Enter the address for the Cloudflare Tunnel configured above in **Value**, for example *nextcloud.example.com*.

{{< trueimage src="/images/SCALE/Apps/NextcloudEnviromentVariables.png" alt="Nextcloud Environment Variables" id="Nextcloud Environment Variables" >}}

### Testing the Cloudflare Tunnel

With the Cloudflare connector and Nextcloud installed and configured, in your Cloudflare dashboard, go to **Networks** and select **Tunnes**.

The status of the tunnel should be **HEALTY**

{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelStateHealthy.png" alt="Cloudflare Tunnel Healty" id="Cloudflare Tunnel Healty" >}}

> Nextcloud should now be reachable via nextcloud.example.com and use a https connection.

Reviewer: ```suggestion
Nextcloud should now be reachable via nextcloud.example.com, using a https connection.
```

### Additional Security Considerations 

> Use strong user passwords and cofigure 2FA for additional security. [Two Factor Authentication](https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/two_factor-auth.html)

Reviewer: ```suggestion
Use strong user passwords and configure [Two Factor Authentication](https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/two_factor-auth.html) for additional security.
```

> Cloudflare offers Access policies to restrict access to the application to specific users, emails or authentication methods. You can find more information in the [applications documentation](https://developers.cloudflare.com/cloudflare-one/applications/).

Reviewer: ```suggestion
Cloudflare offers [access policies](https://developers.cloudflare.com/cloudflare-one/policies/access/) to restrict access to the application to specific users, emails or authentication methods.
```

> Go to **Access**, click **Add an Application** button and select **Self-Hosted** 

Reviewer: ```suggestion
Go to **Access**, click **Add an Application**, and select **Self-Hosted**.
```

> Add your nextcloud application and the domain you configured in the Cloudflare tunnel.

Reviewer: ```suggestion
Add your Nextcloud application and the domain you configured in the Cloudflare tunnel.
```

{{< trueimage src="/images/SCALE/Apps/CloudflareAddApplication.png" alt="Cloudflare Add Application" id="Cloudflare Add Application" >}}

Click **Next**. 

> Create a new new Policy by entering a **Policy Name**. Groups can be assigned to this policy or additional rules can be added. 

Reviewer: ```suggestion
Create a new policy by entering a **Policy Name**. Groups can be assigned to this policy or additional rules can be added. 
```

{{< trueimage src="/images/SCALE/Apps/CloudflareAddPolicy.png" alt="Cloudflare Add Policy" id="Cloudflare Add Policy" >}}

{{< trueimage src="/images/SCALE/Apps/CloudflareCreateAdditionalRules.png" alt="Cloudflare Additional Rules" id="Cloudflare Additional Rules" >}}

Click **Next** and **Save**.

> > Note, that there are additional options for configurations, that are beyond the scope of this tutorial.

Reviewer: ```suggestion
Note: there are additional options for policy configuration, but these are beyond the scope of this tutorial.
```

### Scope

This tutorial provides a general overview of the different options to secure apps. This usually done by installing a additional application client like Cloudflared or Wireguard that proxies traffic form between the user and the application. 
The main focus lies on official apps provided by iXSystems. 
Additionally the user should read the documentation of the app to get a better understanding of the security implications.

Nextcloud will be as an example for this tutorial, the goal is to allow secure access from anywhere.

### Prerequisites

This guide assumes that the user has basic knowledge about TrueNAS Scale itself and is willing to read the [documentation](https://www.truenas.com/docs/scale/scaletutorials/) 

Follow th guide for installing [Nextcloud](https://www.truenas.com/docs/scale/scaletutorials/apps/communityapps/installnextcloudmedia/) on TrueNAS SCALE.

## Cloudflare

For this solution to work, a free Cloudflare account is needed, as well as registered domain. The Cloudflare documentation is very well written and covers everything you need to know, setting up a domain and [DNS](https://developers.cloudflare.com/dns/). 

### Creating a Cloudflare Tunnel

> {{< trueimage src="/images/SCALE/Apps/CloudflareTunnelOverview.jpg" alt="Cloudflare Tunnel Overview" id="Cloudflare Tunnel Overview" >}}
> 
> The Cloudflare Tunnel is a client that runs on the TrueNAS SCALE system and proxies traffic from the Cloudflare network to the application. 

Reviewer: ```suggestion
[Cloudflare Tunnel](https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/) is a client that runs on the TrueNAS SCALE system and proxies traffic from the Cloudflare network to the application.

{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelOverview.jpg" alt="Cloudflare Tunnel Overview" id="Cloudflare Tunnel Overview" >}}

```

Reviewer: Note we should add a credit to Cloudflare for the overview image

> It is highly recommended to read the [tunnel documentation](https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/) or to watch [this](https://www.youtube.com/watch?v=eojWaJQvqiw) from Lawrence Systems setting up Cloudflare tunnels for applications. This guide assumes that the applications run as a docker container but the same approach can be used to secure apps running on Truenas Scale in Kubernetes.

Reviewer: ```suggestion
This guide assumes that the applications run as a docker container but the same approach can be used to secure apps running on Truenas Scale in Kubernetes.
 [This video](https://www.youtube.com/watch?v=eojWaJQvqiw) from Lawrence Systems provides a detailed overview of setting up Cloudflare tunnels for applications. 
```

> In the Cloudflare One dashboard, set a public hostname for accessing Nextcloud, this hostname, in this example it is:  **nextcloud.example.com**

Reviewer: ```suggestion
In the Cloudflare One dashboard, set a public hostname for accessing Nextcloud, for example:  *nextcloud.example.com*.
```

> {{< trueimage src="/images/SCALE/Apps/CloudflareTunnelHostname.png" alt="Cloudflare Tunnel Hostname" id="Cloudflare Tunnel Hostname" >}}
> 
> The service Type is set to **HTTPS** and the URL is the local TrueNAS IP + the Nextcloud container port, e.g. 192.168.1.1:9001

Reviewer: ```suggestion
Set service **Type** to **HTTPS**.
Enter the local TrueNAS IP with the Nextcloud container port, for example *192.168.1.1:9001*.

{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelHostname.png" alt="Cloudflare Tunnel Hostname" id="Cloudflare Tunnel Hostname" >}}

```

> {{< trueimage src="/images/SCALE/Apps/CloudflareTunnelTLSSetting.png" alt="Cloudflare Tunnel TLS Setting" id="Cloudflare Tunnel TLS Setting" >}}
> 
> Additional application Settings > TLS > No TLS Verify = Enabled

Reviewer: ```suggestion
Go to **Additional application Settings**, select **TLS** from the dropdown menu, and enable **No TLS Verify**.

{{< trueimage src="/images/SCALE/Apps/CloudflareTunnelTLSSetting.png" alt="Cloudflare Tunnel TLS Setting" id="Cloudflare Tunnel TLS Setting" >}}

```

### Nextcloud Configuration

> Two environment variables must be set in the Nextcloud application.
> [overwrite.cli.url](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwrite-cli-url) and [overwritehost](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwritehost)

Reviewer: ```suggestion
Install the [Nextcloud]({{< relref "/scale/scaletutorials/apps/communityapps/installnextcloudmedia.md" >}}) community application.

Click **Add** next to **Nextcloud environment** to configure the environment variables [overwrite.cli.url](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwrite-cli-url) and [overwritehost](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#overwritehost).
```

> In the Nextcloud App the environment variables are edited as shown below, where `value` is the address. 

Reviewer: ```suggestion
Enter the two environment variables in **Name** as *OVERWRITECLIURL* and *OVERWRITEHOST*.

Enter the address for the Cloudflare Tunnel configured above in **Value**, for example *nextcloud.example.com*.

```

> {{< trueimage src="/images/SCALE/Apps/NextcloudEnviromentVariables.png" alt="Nextcloud Environment Variables" id="Nextcloud Environment Variables" >}}

Reviewer: At the end can we wrap up the procedure with a step for confirming the configuration succeeded? Maybe something like wait for the installed application to deploy and then access via the proxy URL?
